pkg/scalers/temporal_scaler.go

```diff
@@ -0,0 +1,269 @@
+			},
+		},
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Temporal client: %w", err)
+	}
+
+	return &temporalWorkflowScaler{
+		metricType: metricType,
+		metadata:   meta,
+		tcl:        c,
+		logger:     logger,
+	}, nil
+}
+
+// Close closes the Temporal client connection.
+func (s *temporalWorkflowScaler) Close(context.Context) error {
+	if s.tcl != nil {
+		s.tcl.Close()
+	}
+	return nil
+}
+
+// GetMetricSpecForScaling returns the metric specification for scaling.
+func (s *temporalWorkflowScaler) GetMetricSpecForScaling(context.Context) []v2.MetricSpec {
+	externalMetric := &v2.ExternalMetricSource{
+		Metric: v2.MetricIdentifier{
+			Name: s.metadata.metricName,
+		},
+		Target: GetMetricTarget(s.metricType, s.metadata.targetQueueSize),
+	}
+	metricSpec := v2.MetricSpec{
+		External: externalMetric,
+		Type:     externalMetricType,
+	}
+	return []v2.MetricSpec{metricSpec}
+}
+
+// GetMetricsAndActivity returns metrics and activity for the scaler.
+func (s *temporalWorkflowScaler) GetMetricsAndActivity(ctx context.Context, metricName string) ([]external_metrics.ExternalMetricValue, bool, error) {
+	queueSize, err := s.getQueueSize(ctx)
+	if err != nil {
+		return nil, false, fmt.Errorf("failed to get Temporal queue size: %w", err)
+	}
+
+	metric := GenerateMetricInMili(metricName, float64(queueSize))
+
+	return []external_metrics.ExternalMetricValue{metric}, queueSize > s.metadata.activationTargetWorkflowLength, nil
+}
+
+// getQueueSize returns the queue size of open workflows.
+func (s *temporalWorkflowScaler) getQueueSize(ctx context.Context) (int64, error) {
+
+	var executions []executionInfo
+	var nextPageToken []byte
+	for {
+		listOpenWorkflowExecutionsRequest := &workflowservice.ListOpenWorkflowExecutionsRequest{
+			Namespace:       s.metadata.namespace,
+			MaximumPageSize: 1000,
+			NextPageToken:   nextPageToken,
+			Filters: &workflowservice.ListOpenWorkflowExecutionsRequest_TypeFilter{
+				TypeFilter: &tclfilter.WorkflowTypeFilter{
+					Name: s.metadata.workflowName,
+				},
+			},
+		}
+		ws, err := s.tcl.ListOpenWorkflow(ctx, listOpenWorkflowExecutionsRequest)
+		if err != nil {
+			return 0, fmt.Errorf("failed to get workflows: %w", err)
+		}
+
+		for _, exec := range ws.GetExecutions() {
+			execution := executionInfo{
+				workflowId: exec.Execution.GetWorkflowId(),
+				runId:      exec.Execution.RunId,
+			}
+			executions = append(executions, execution)
+		}
+
+		if nextPageToken = ws.NextPageToken; len(nextPageToken) == 0 {
+			break
+		}
+	}
+
+	pendingCh := make(chan string, len(executions))
+	var wg sync.WaitGroup
+
+	for _, execInfo := range executions {
+		wg.Add(1)
+		go func(e executionInfo) {
+			defer wg.Done()
+
+			workflowId := e.workflowId
+			runId := e.runId
+
+			if !s.isActivityRunning(ctx, workflowId, runId) {
+				executionId := workflowId + "__" + runId
+				pendingCh <- executionId
+			}
+
+		}(execInfo)
+	}
+	wg.Wait()
+	close(pendingCh)
```

Comment:
Same as above, if there are thousand of pending executions, What will happen?
Probably I'm wrong, but I understand that we are listing all the workflows and inside the workflows, we are checking all the executions to decide if it's running or not. This suggests me some questions:
- Are executed activities removed at any moment or will we have this queue growing and growing?
- Doesn't the workflow have any option to give that information during the first requests instead of having to navigate over all the activities?